doc_internal/fish-riir-plan.md

```diff
@@ -0,0 +1,77 @@
+  - Go would be an awkward fit. fork is [quite the problem](https://stackoverflow.com/questions/28370646/how-do-i-fork-a-go-process/28371586#28371586) in Go.
+  - Other system languages (D, Nim, Zig...) are too niche: fewer contributors, higher risk of the language becoming irrelevant.
+
+## Risks
+
+- Large amount of work with possible introduction of new bugs.
+- Long period of complicated builds.
+- Existing contributors will have to learn Rust.
+- As of yet unknown compatibility story for Tier 2+ platforms (Cygwin, etc).
+
+## Approach
+
+We will do an **incremental port** in the span of one release. We will have a period of using both C++ and Rust, and both cargo and CMake, leveraging FFI tools (see below).
+
+The work will **proceed on master**: no long-lived branches. Tests and CI continue to pass at every commit for recent Linux and Mac. Centos7, \*BSD, etc may be temporarily disabled if they prove problematic.
+
+The Rust code will initially resemble the replaced C++. Fidelity to existing code is more important than Rust idiomaticity, to aid review and bisecting. But don't take this to extremes - use judgement.
+
+The port will proceed "outside in." We'll start with leaf components (e.g. builtins) and proceed towards the core. Some components will have both a Rust and C++ implementation (e.g. FLOG), in other cases we'll change the existing C++ to invoke the new Rust implementations (builtins).
+
+After porting the C++, we'll replace CMake.
+
+We will continue to use wide chars, locales, gettext, printf format strings, and PCRE2. We will not change the fish scripting language at all. We will _not_ use this as an opportunity to fix existing design flaws, with a few carefully chosen exceptions. See [Strings](#strings).
+
+We will not use tokio, serde, async, or other fancy Rust frameworks initially.
+
+### FFI
+
+Rust/C++ interop will use [autocxx](https://github.com/google/autocxx), [Cxx](https://cxx.rs), and possibly [bindgen](https://rust-lang.github.io/rust-bindgen/). I've forked these for fish (see the Development Guide). Once the port is done, we will stop using them, except perhaps bindgen for PCRE2.
+
+We will use [corrosion](https://github.com/corrosion-rs/corrosion) for CMake integration.
+
+Inefficiencies (e.g. extra string copying) at the FFI layer are fine, since it will all get thrown away.
+
+Tests can stay in fish_tests.cpp or be moved into Rust .rs files; either is fine.
+
+### Strings
+
+Rust's `String` / `&str` types cannot represent non-UTF8 filenames or data using the default encoding scheme. That's why all string conversions must go through fish's encoding scheme (using the private-use area to encode invalid sequences). For example, fish cannot use `File::open` with a `&str` because the decoding will be incorrect.
+
+So instead of `String`, fish will use its own string type, and manage encoding and decoding as it does today. However we will make some specific changes:
```

Comment:
Could [OsString](https://doc.rust-lang.org/std/ffi/struct.OsString.html) be used?

Reply:
It would be confusing to use it, since `OsStr` on \*nix is generally byte-oriented while fish's strings are currently Unicode-32.

Reply:
You mean UTF-32. There’s no such thing called Unicode-32.